src/transformers/utils/generic.py

```diff
@@ -926,29 +955,140 @@ def can_return_tuple(func):
+# if is_torch_available():
+# @torch._dynamo.disable
+@dataclass
+@requires(backends=("torch",))
+class OutputRecorder:
+    """
+    Configuration for recording outputs from a model via hooks.
 
-        return output
+    Attributes:
+        target_class (Type): The class (e.g., nn.Module) to which the hook will be attached.
+        index (Optional[int]): If the output is a tuple/list, optionally record only at a specific index.
+        layer_name (Optional[str]): Name of the submodule to target (if needed), e.g., "transformer.layer.3.attn".
+    """
+
+    target_class: "type[torch.nn.Module]"
+    index: Optional[int] = 0
+    layer_name: Optional[str] = None
+
+
+def check_model_inputs(func):
+    """
+    Decorator to intercept specific layer outputs without using hooks.
+    Compatible with torch.compile (Dynamo tracing).
+    """
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        use_cache = kwargs.get("use_cache", getattr(self.config, "use_cache", False))
+        return_dict = kwargs.pop("return_dict", getattr(self.config, "return_dict", True))
+        all_args = kwargs.copy()
+
+        if getattr(self, "gradient_checkpointing", False) and self.training and use_cache:
+            logger.warning_once(
+                "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`."
+            )
+            kwargs["use_cache"] = False
+
+        if "kwargs" in all_args:
+            for k, v in all_args["kwargs"].items():
+                all_args[k] = v
+
+        capture_flags = _CAN_RECORD_REGISTRY[self] or []  # there is a weak ref for executorch
+        recordable_keys = {
+            f"output_{k}": all_args.get(
+                f"output_{k}",
+                getattr(
+                    self.config,
+                    f"output_{k}",
+                    all_args.get("output_attentions", getattr(self.config, "output_attentions", False)),
+                ),
+            )
+            for k in capture_flags
+        }
+        collected_outputs = defaultdict(tuple)
+        monkey_patched_layers = []
+
+        def make_capture_wrapper(module, orig_forward, key, index):
+            @wraps(orig_forward)
+            def wrapped_forward(*args, **kwargs):
+                output = orig_forward(*args, **kwargs)
+                if not isinstance(output, tuple):
+                    collected_outputs[key] += (output,)
+                elif output[index] is not None:
+                    collected_outputs[key] += (output[index],)
+                # Manually invoke forward hooks (preserve original hook behavior)
+                if module._forward_hooks:
+                    for hook in module._forward_hooks.values():
+                        hook_result = hook(module, args, output)
+                        if hook_result is not None:
+                            output = hook_result
+                return output
```

Comment:
hooks are already run in the orig_forward
```suggestion
            def wrapped_forward(*args, **kwargs):
                output = orig_forward(*args, **kwargs)
                if not isinstance(output, tuple):
                    collected_outputs[key] += (output,)
                elif output[index] is not None:
                    collected_outputs[key] += (output[index],)
                return output
```